pkg/agent/proxy/meta_proxier.go

```diff
@@ -0,0 +1,182 @@
+// Copyright 2020 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package proxy
+
+import (
+	"fmt"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/klog"
+	utilnet "k8s.io/utils/net"
+
+	"github.com/vmware-tanzu/antrea/pkg/agent/proxy/types"
+	k8sproxy "github.com/vmware-tanzu/antrea/third_party/proxy"
+)
+
+type metaProxier struct {
+	ipv4Proxier types.Provider
+	ipv6Proxier types.Provider
+}
+
+// NewMetaProxier returns a dual-stack "meta-proxier". Proxier API
+// calls will be dispatched to the ProxyProvider instances depending
+// on address family.
+func NewMetaProxier(ipv4Proxier, ipv6Proxier types.Provider) types.Provider {
+	return types.Provider(&metaProxier{
+		ipv4Proxier: ipv4Proxier,
+		ipv6Proxier: ipv6Proxier,
+	})
+}
+
+// SyncLoop runs periodic work.  This is expected to run as a
+// goroutine or as the main loop of the app.  It does not return.
+func (proxier *metaProxier) SyncLoop() {
+	go proxier.ipv6Proxier.SyncLoop() // Use go-routine here!
```

Comment:
Why use goroutine for ipv6Proxier, but not for ipv4Proxier?

Reply:
SyncLoop() itself is called using goroutine. So ipv4proxier resume in parent context (and runs indefinitely), while ipv6proxier spawns from a new context.

Reply:
In my opinion, the process for IPv4 and IPv6 should be similar. If leaving ipv4Proxier in the parent context just to ensure the function to run infinitely, we could use a waitGroup. No strong preferrence.

Reply:
Using a waitGroup will retain one extra context which will pend forever as SyncLoop runs indefinitely, right?